### PR TITLE
[k8s] Device and hubname can start with numerical values.

### PR DESCRIPTION
**Pull request by darobs**
Make device and hubname selector and labels all match

### CODE REVIEW

File: edgelet/edgelet-kube/src/convert/mod.rs [hunk 1]
```diff
@@ -13,21 +13,6 @@ pub use to_k8s::{
...
-pub fn sanitize_dns_domain(domain: &str) -> Result<String> {
-    let sanitized = domain
-        .split('.')
-        .map(|label| sanitize_dns_value(label))
-        .collect::<Result<Vec<String>>>()?
-        .join(".");
-    if sanitized.len() <= DNS_DOMAIN_MAX_SIZE {
-        Ok(sanitized)
-    } else {
-        Err(ErrorKind::InvalidDnsName(domain.to_owned()).into())
```
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/mod.rs]: Probably not, will check.

File: edgelet/edgelet-kube/src/convert/mod.rs [hunk 2]
```diff
@@ -37,6 +22,23 @@ pub fn sanitize_dns_value(name: &str) -> Result<String> {
...
 }
 
+// Valid label values must be 63 characters or less and must be empty or begin and end with an
+// alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between.
+const LABEL_MAX_SIZE: usize = 63;
+fn is_allowed_label_chars(c: char) -> bool {
+    c.is_ascii_alphanumeric() || c == '_' || c == '-' || c == '.'
+}
+
+pub fn sanitize_label_value(name: &str) -> String {
```
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/mod.rs]: Maybe? Can a regex in Rust be static? I wouldn't want to create a new regex every time I invoke `sanitize_label_value`.

[the remark this answers is not in the document]
Comment (dmolokanov) [reply edgelet/edgelet-kube/src/convert/mod.rs]: Of course. You can define in inside lazy_static block. Like [here](https://docs.rs/regex/1.3.5/regex/#example-avoid-compiling-the-same-regex-in-a-loop)

[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/mod.rs]: Yes, empty values are valid. However, edgeAgent doesn't like them.  I've been debating back and forth to have this return an error if the device (or hubname) winds up empty. I settled on edgeAgent crashing.

File: edgelet/edgelet-kube/src/convert/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/mod.rs]: Removed.

File: edgelet/edgelet-kube/src/convert/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/mod.rs]: I didn't like any regex implementation.  Best I got was:
```Rust
pub fn sanitize_label_value(name: &str) -> String {
 lazy_static! {
 static ref RE: Regex = Regex::new(r"[^[:alnum:]_\-.]").unwrap();
 }
 
 let mut picked = RE.replace_all(name, "").into_owned();
 picked.truncate(LABEL_MAX_SIZE);
 picked
 .trim_start_matches(|c: char| !c.is_ascii_alphanumeric())
 .trim_end_matches(|c: char| !c.is_ascii_alphanumeric())
 .to_lowercase()
}
```

Which isn't any more readable or understandable to me.  It did make me add test cases, and I tweaked the sanitize function to meet all the new tests.

File: edgelet/edgelet-kube/src/convert/mod.rs [hunk 3]
```diff
@@ -31,12 +31,12 @@ fn is_allowed_label_chars(c: char) -> bool {
 
 pub fn sanitize_label_value(name: &str) -> String {
     name.trim_start_matches(|c: char| !c.is_ascii_alphanumeric())
-        .trim_end_matches(|c: char| !c.is_ascii_alphanumeric())
-        .to_lowercase()
         .chars()
         .filter(|c| is_allowed_label_chars(*c))
         .take(LABEL_MAX_SIZE)
         .collect::<String>()
```
[the remark this answers is not in the document]
Comment (darobs) [reply edgelet/edgelet-kube/src/convert/mod.rs]: Still had to trim the end after the truncate. (See test case on line 107).  I only create a new string if I hit that edge case.
